[v10] Add in app access connection monitoring.

mdwn opened the pull request:
Backport https://github.com/gravitational/teleport/pull/17436 to branch/v10

mdwn committed:
Add in app access connection monitoring.

Application access connection monitoring has been introduced so that, when a
lock is created, application access connections will be interrupted until the
lock has been cleared. This includes web sockets and TCP applications.

r0mant reviewed (dismissed):
Do not merge until new tests flakiness is fixed.

mdwn committed:
Make TestTCP* tests in appaccess more deterministic. (#18233)

The appaccess TestTCP* tests are highly reliant on time. This has been
reduced (but not eliminated) by using a fakeClock and a channel for
signaling monitor triggered connection closures.
